---

[//]: # "Header"

<p style="text-align: left;  font-size:18pt; LINE-HEIGHT:30px;">
    <span style="float: left">
     Technische Hochschule Ingolstadt<br>
     Prof. Dr. Sören Gröttrup <br>
     Laura Dietl
    </span>
    <span style="float: right;">
       Machine Learning 1<br>
        <span style="float: right;">WS 24/25</span>
    </span>
</p>

[//]: # "Header Aufgabenblatt"
<br>
<p style="text-align: center;  font-size:18pt; LINE-HEIGHT:30px;">
     <span style="font-weight: bold;">Aufgabenblatt 4</span><br>
     Themen: Imputing, Neuronale Netze in PyTorch, Backpropagation<br>
     Abgabetermin: 29.11.2024, 23:59 Uhr <br>
     Punkte: 31
</p>


---

**Namen:** Emelie Hauck, Conrad Christoph 

---

# 1. Imputing - Dealing with Missing Values - Ein Weg mit fehlenden Werten umzugehen

Bei einer Variablen in einem Datensatz kann es durchaus vorkommen, dass für einige Beobachtungen die Werte nicht bekannt sind. Zum Beispiel kann es sein, dass ein Kunde in einem Online-Shop sein Profil nicht vollkommen ausfüllt, so dass die Handynummer, das Alter oder gewisse Interessen nicht bekannt sind. Die Variable Handynummer in diesem Datensatz wäre dann für den jeweiligen Kunden nicht befüllt.

Einige Algorithmen des Maschinellen Lernens können mit diesen fehlenden Werten nicht umgehen. Damit man einen Algorithmus trotzdem auf einen Datensatz mit einer Variable mit fehlenden Werten anwenden kann, gibt es einige einfache Strategien.

**Ignoriere die fehlenden Werte (Omission)**

Die einfachste und radikalste Methode ist die fehlenden Werte für das Training zu entfernen. Dies kann man auf zwei Arten machen:
1. Entfernen der Variable aus dem Trainingsdatensatz. Z.B. im obigen Beispiel die Variable Handynummer nicht als Input-Variable zu nehmen.
1. Entfernen aller Zeilen im Datensatz, die einen fehlenden Wert in einer zu berücksichtigen Variable aufweisen. Z.B. im obigen Beispiel alle Kunden ohne Handynummer aus dem Tainingsdatensatz ausschließen.

**Imputing**

Anstatt die fehlenden Werte zu entfernen, kann man auch versuchen, diese mit "geeigneten" Werten aufzufüllen. Dafür gibt es mehrere einfache Methoden.
1. _Constant Value Imputing_ Wenn der Kontext es hergibt, können alle fehlenden Werte einer Variablen mit einem Standardwert aufgefüllt werden. 
    1. Bei einer kategoriellen Variable können z.B. die fehlenden Werte mit 'unknown' oder 'other' ersetzt werden. Wenn z.B. bei einem Kunden der Wert der Variable "Intresse an Sportartikeln" fehlt, kann dieser die neue Ausprägung "unknown" erhalten.
    1. Wenn bei einer numerischen Variablen nur positive Werte auftreten können, wie z.B. bei der Variable "Anzahl an Käufe", kann der fehlende Wert sinnvollerweise mit 0 ersetzt werden.
1. _Mean/Median/Mode Imputing_ Die fehlenden Werte einer Variablen werden durch deren arithmetisches Mittel, den Median oder den Modus ersetzt. Diese repräsentiern "typische" Werte für die Variable.
    1. Bei kategoriellen Variablen werden die fehlenden Werte durch die am häufigsten auftretende Ausprägung ersetzt (den Modus).
    1. Bei numerischen Variablen werden die fehlenden Werte mit dem arithmetischen Mittel oder dem Median der anderen Ausprägungen der Variable ersetzt.

Dies sind nur die einfachen und univariaten Methoden des Imputing. Weitere multivariate Imputing-Verfahren wie die Benutzung von Vorhersagemodellen werden in der Vorlesung besprochen.

## Aufgabe 1.1 (Vor- & Nachteile der Imputing-Verfahren) [6 Punkte]
Betrachten Sie die oben genannten Omission und Imputing-Verfahren. Welche Vor- und Nachteile könnten diese Verfahren jeweils haben?

**Antwort zu Aufgabe 1.1**


Omission:
| Verfahren | Vorteile             | Nachteile                                                                                              |
|-----------|----------------------|--------------------------------------------------------------------------------------------------------|
| 1.        |- Alle Daten sind Uniform <br>- eingefügte Werte ändern die Ergebnisse nicht/keine nicht vorhandenen Daten müssen angenommen werden <br>- alle Datenpunkte bleiben erhalten <br > - unkompliziertes/schnelles Verfahren <br> - gut bei wenig fehlenden Werten |- Die entfernte Variable ist potentiell ausschlaggebend für das Ergebnis <br>                             |
| 2.        |- Alle Daten sind Uniform <br>- eingefügte Werte ändern die Ergebnisse nicht/keine nicht vorhandenen Daten müssen angenommen werden <br>- potentiell ausschlaggebende Variablen bleiben erhalten  <br> - gut bei wenig fehlenden Werten                         |- Sollten viele Datenpunkte unvolsständig sein, so führt das Verfahren zu großem Verlust knapper Daten   |

Imputing:
| Verfahren                      | Vorteile                                                                                                                                     | Nachteile |
|--------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------|-----------|
| Constant Value Imputing        |- keine nicht vorhandenen Angaben werden angenommen<br>- Struktur der Daten bleibt erhalten <br>- Stichprobenumfang bleibt erhalten           |- Das Modell könnte Ergebnisse o. zu viel Information aus der Unknown Variable ziehen<br> - Es werden nicht korrekte Daten angenommen, die die Vorhersage verzerren könnten                                                |
| Mean/Median/Mode Imputing      |- kann über verschiedene Annahmemethoden (Mean,Median,etc.) Strukturen im Datensatz berücksichtigen<br>- Stichprobenumfang bleibt erhalten    | - basiert auf Annahmen, die teilweise aufwändig errechnet werden müssen<br>- es könnten systematische Verzerrungen entstehen                                                                                            | 


# 2. Neuronale Netze mit PyTorch

### Daten

Der Datensatz `gt.csv` (zu finden auf der Moodle-Seite des Praktikums (https://moodle.thi.de/course/view.php?id=6824)) ist eine etwas abgewandelte und bearbeitete Version des Datensatzes aus dem UCI Machine Learning Repository (https://archive.ics.uci.edu/ml/datasets/Gas+Turbine+CO+and+NOx+Emission+Data+Set). 

Der Datensatz enthält 36733 Messpunkte von je 11 Sensoren einer Gasturbine im Nordwesten der Türkei. Die einzelnen Werte sind dabei jeweils Aggregate (Summe oder Mittelwert) über eine Stunde. Die Sensordaten diesen dazu die Gasemissionender Turbine, insb. CO und NOx (NO & NO2), zu messen.

#### Initialisierung der Daten
Hier finden Sie bereits geschriebenen Code, der die Daten einliest und in Trainings- und Testdaten splittet. __Bitte verändern Sie diesen Code nicht__, sondern fügen Ihren Code unterhalb ein.

In [1]:
# Packages
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split

plt.style.use('ggplot')

In [2]:
# Load data
file_path = "Daten Blatt 4/gt.csv"
data = pd.read_csv(file_path)
data

AT      AP      AH    AFDP    GTEP     TIT     TAT     TEY     CDP  \
0      4.5878  1018.7  83.675  3.5758  23.979  1086.2  549.83  134.67     NaN   
1      4.2932  1018.3  84.235  3.5709  23.951  1086.1     NaN  134.67  11.892   
2      3.9045  1018.4  84.858  3.5828  23.990  1086.5  550.19  135.10  12.042   
3      3.7436  1018.3  85.434  3.5808  23.911  1086.5  550.17  135.03  11.990   
4      3.7516  1017.8  85.182  3.5781  23.917  1085.9  550.00  134.67  11.910   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
36728  3.6268  1028.5  93.200  3.1661  19.087  1037.0  541.59  109.08  10.411   
36729  4.1674  1028.6  94.036  3.1923  19.016  1037.6  542.28  108.79  10.344   
36730  5.4820  1028.5  95.219  3.3128  18.857  1038.0  543.48  107.81  10.462   
36731  5.8837  1028.7  94.200  3.9831  23.563  1076.9  550.11  131.41  11.771   
36732  6.0392  1028.8  94.547  3.8752  22.524  1067.9  548.23  125.41  11.462   

             CO      NOX  
0       0.32663   81.952  
1       0.44784   82.377  
2       0.45144   83.776  
3       0.23107   82.505  
4       0.26747   82.028  
...         ...      ...  
36728  10.99300   89.172  
36729  11.14400   88.849  
36730  11.41400   96.147  
36731   3.31340   64.738  
36732  11.98100  109.240  

[36733 rows x 11 columns]

In [3]:
# Define target and feature
y = data.NOX
X = data.drop(columns=["CO", "NOX"])

In [4]:
# Split data into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=1)

X_train = pd.DataFrame(X_train, columns=X.columns)
X_test = pd.DataFrame(X_test, columns=X.columns)

## Aufgabe 2.1 (Datenaufbereitung) [4 Punkte]

Untersuchen Sie den Datensatz der Sensordaten einer Gasturbine `gt.csv`.

1. Welche Variablen weisen fehlende Werte auf?
1. Ersetzen Sie diese mittels eines sinnvollen oben genannten Imputing-Verfahrens. _Hinweis_: Sie können dafür die Funktion `SimpleImputer()` aus dem `sklearn` Paket verwenden. Informieren Sie sich vorher über deren Funktionsweise und Optionen.
1. Was für ein Vorverarbeitungsschritt scheint noch sinnvoll zu sein? Führen Sie diesen ggf. auch durch.

_Achtung_: Behalten Sie bei Ihren Datenoperationen die gegebene Aufteilung in Training- und Testdaten bei.

### Lösung Aufgabe 2.1

In [5]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

In [6]:
# 2.1.1

empty_cells = data.isnull().any(axis=0)
print(empty_cells)

AT      False
AP      False
AH      False
AFDP    False
GTEP    False
TIT     False
TAT      True
TEY     False
CDP      True
CO      False
NOX     False
dtype: bool


Die Variablen TAT & CDP weisen fehlende Werte auf.

In [7]:
# 2.1.2

imputer = SimpleImputer(strategy='median')

imputer.fit(X_train)
imputed_x_train = imputer.transform(X_train)

imputer.fit(X_test)
imputed_x_test = imputer.transform(X_test)

#### 2.1.3

Die Daten weisen unterschiedliche Größen auf (z.B. AT im 1-2 stelligen Bereich vor dem Komma und AP im vierstelligen Bereich vor dem Komma) entsprechend wäre eine Skalierung sinnvoll, da einige Verfahren nur mit skalierten Daten präzise Ergebnisse liefern.    

In [8]:
scaler = StandardScaler()

x_train_sc_imp = scaler.fit_transform(imputed_x_train)

x_test_sc_imp = scaler.transform(imputed_x_test)

## Aufgabe 2.2 (Training in PyTorch) [12 Punkte]
Erstellen Sie ein Feedforward Neuronales Netz auf dem in der vorherigen Aufgabe 2.1 aufbereiteten Datensatz mittels PyTorch, welches basierend auf den anderen Sensordaten die NOx-Emission vorhersagen kann. Benutzen Sie dabei den L2-Loss als Kostenfunktion. Welchen RMSE erzielt Ihr trainiertes Neuronales Netz, wenn Sie dies auf die Testdaten (`X_test, y_test`) anwenden?

In dieser Aufgabe gibt es keinerlei weitere Vorgabe an das zu implementierende Modell. Sie sind in der Struktur des Netzes und dem Vorgehen frei. Um aber die volle Punkzahl zu erhalten, müssen Sie ein sinnvolles Netz trainieren, d.h. die Optimierung mittels Gradient Descent konvergiert, kein Overfitting etc.. Versuchen Sie auch den RMSE so klein wie möglich zu bekommen.

_Hinweis_: Anstatt Ihre aufbereiteten Daten zu nehmen, können Sie auch schon vorverarbeitete Daten einlesen durch den Befehl `data = torch.load("Daten Blatt 4/gt_prepared")`
Dieser liest ein Dictionary ein, in welchem Sie Tensoren der Trainings- und Testdaten finden. Der Befehl `data.keys()` zeigt Ihnen die Key-Namen. 

### Lösung der Aufgabe 2.2

In [9]:
# Load packages
import torch
import torch.nn as net
import torch.nn.functional as F
import torch.optim as optim


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.1.3 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/emelie/Library/Python/3.12/lib/python/site-packages/ipykernel_launcher.py", line 18, in <module>
    app.launch_new_instance()
  File "/Users/emelie/Library/Python/3.12/lib/python/site-packages/traitlets/config/application.py", line 1075, in launch_instance
    app.start()
  File "/Users/emelie/Library/Python/3.12/lib/python/site-packages/ipykernel/kernelapp.py", line 739, in start
    self.io_loop.start(

In [12]:
data = torch.load("Daten Blatt 4/gt_prepared")
data.keys()

X_train_new = data["train_data"]
y_train_new = data["train_target"]
y_train_new = y_train_new.unsqueeze(1)

X_test_new = data["test_data"]
y_test_new = data["test_target"]
y_test_new = y_test_new.unsqueeze(1)

In [13]:
class FeedforwardNeuralNet(net.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(FeedforwardNeuralNet, self).__init__()
        self.neuralnet = net.Sequential(
            net.Linear(input_size, hidden_size),
            net.ReLU(),                     #ReLU als Aktivierungsfunktion
            net.Linear(hidden_size,32),
            net.ReLU(),
            net.Linear(32,output_size)
        ) 
    
    def forward(self, x):
        return self.neuralnet(x)

if len(X_train_new.shape) == 1:
    X_train_new = X_train_new.unsqueeze(1)
if len(X_test_new.shape) == 1:
    X_test_new = X_test_new.unsqueeze(1)

input_size = X_train_new.shape[1]

hidden_size = 64
output_size = 1
alpha = 0.01
epochs = 1000

model = FeedforwardNeuralNet(input_size, hidden_size, output_size)

loss = net.MSELoss()    
optimizer = optim.Adam(model.parameters(), lr=alpha)

#Training
for epoch in range(epochs):
    optimizer.zero_grad()
    y_pred = model(X_train_new)

    train_loss = loss(y_pred, y_train_new)
    train_loss.backward()
    optimizer.step()
    print(f"Epoch {epoch+1}/{epochs}, Loss: {train_loss.item()}")

#Test RMSE
model.eval()
with torch.no_grad():
    test_predictions = model(X_test_new)
    test_mse = loss(test_predictions, y_test_new)
    test_rmse = torch.sqrt(test_mse)

print(f"RMSE auf Testdaten: {test_rmse.item():.4f}")

#Trainings-RMSE letzte Iteration
final_train_loss = train_loss.item() 
final_train_rmse = final_train_loss ** 0.5  
print(f"Finaler Trainings-Loss (MSE): {final_train_loss:.4f}")
print(f"Finaler Trainings-RMSE: {final_train_rmse:.4f}")



Epoch 1/1000, Loss: 4409.2314453125
Epoch 2/1000, Loss: 4372.33251953125
Epoch 3/1000, Loss: 4332.87109375
Epoch 4/1000, Loss: 4286.79248046875
Epoch 5/1000, Loss: 4230.49072265625
Epoch 6/1000, Loss: 4160.7607421875
Epoch 7/1000, Loss: 4075.4658203125
Epoch 8/1000, Loss: 3972.336181640625
Epoch 9/1000, Loss: 3849.455078125
Epoch 10/1000, Loss: 3705.02197265625
Epoch 11/1000, Loss: 3537.674560546875
Epoch 12/1000, Loss: 3346.65380859375
Epoch 13/1000, Loss: 3131.988525390625
Epoch 14/1000, Loss: 2894.736572265625
Epoch 15/1000, Loss: 2637.3232421875
Epoch 16/1000, Loss: 2364.01171875
Epoch 17/1000, Loss: 2081.348876953125
Epoch 18/1000, Loss: 1798.6314697265625
Epoch 19/1000, Loss: 1528.115478515625
Epoch 20/1000, Loss: 1284.858642578125
Epoch 21/1000, Loss: 1085.757080078125
Epoch 22/1000, Loss: 946.7763061523438
Epoch 23/1000, Loss: 877.3410034179688
Epoch 24/1000, Loss: 872.9905395507812
Epoch 25/1000, Loss: 911.2461547851562
Epoch 26/1000, Loss: 956.3572387695312
Epoch 27/1000, Los

Da der RMSE auf den Testdaten und der RMSE auf den Trainingsdaten (nach der letzten Iteration) sehr ähnlich sind, kann man Overfitting ausschließen.

# 3. Parameter eines Neuronalen Netzes

## Aufgabe 3.1 [3 Punkte]
Gegeben ist das in PyTorch definierte Neuronale Netze
   
![fc_nn](Bilder/fc_nn.png)

Bestimmen Sie  die Anzahl zu optimierender Parameter.

### Antwort zu Aufgabe 3.1

Anzahl zu optimierender Parameter = Gewichte und Biases für jede Schicht

Eingabedimension (input_size) = 9 <br>

Schichten: <br>
fc1: 9 Eingaben, 8 Ausgaben <br>
fc2: 8 Eingaben, 6 Ausgaben <br>
fc3: 6 Eingaben, 1 Ausgabe <br>

**fc1** (Linear(9, 8)): <br>
Gewichte: 9 × 8 = 72 <br>
Biases: 8 <br>
Gesamt: 72 + 8 = 80 <br>

**fc2** (Linear(8, 6)): <br>
Gewichte: 8 × 6 = 48 <br>
Biases: 6 <br>
Gesamt: 48 + 6 = 54 <br>

**fc3** (Linear(6, 1)): <br>
Gewichte: 6 × 1 = 6 <br>
Biases: 1 <br>
Gesamt: 6 + 1 = 7 <br>

**Gesamtzahl der Parameter:** <br>
80 + 54 + 7 = 141

In der Anwendung kämen als Parameter noch die Learning Rate alpha und die Anzahl der Epochen dazu.

# 4. Backpropagation

## Aufgabe 4.1 [6 Punkte]
Gegeben ist folgendes Neuronale Netze
   
![fc_nn](Bilder/NN_backpropagation.png)

mit folgenden Gewichtsmatrizen 

$$W^{(1)}=\left(\begin{array}{cc} 
1 & 1\\ 
2 & -2
\end{array}\right)
\quad\text{und}\quad
W^{(2)}=\left(\begin{array}{c} 
1\\ 
-1
\end{array}\right) $$

und Biases

$$b^{(1)}=\left(\begin{array}{c} 
-1\\ 
0
\end{array}\right)
\quad\text{und}\quad
b^{(2)}= -4$$

Die Aktivierungsfunktionen sind folgende:
* Hidden Layer: ReLU
* Output Layer: Identität

Als Verlustfunktion wird der L2-Loss genommen
$$L(y,a)=\frac12(y-a)^2$$

__Aufgabe:__
1. Führen Sie ein Update der Parameter mittels Backpropagation durch für den Datenpunkt
$$x=\left(\begin{array}{c} 
1\\ 
1 
\end{array}\right) $$
mit wahrem Zielwert 
$$ y= -1$$
und der Lernrate $\alpha=0.1$ durch.
1. Wie groß ist der neue Verlust nach dem Parameterupdate? Vergleichen Sie das mit dem Verlust vor durchführung des Updates der Parameter.


### Lösung Aufgabe 4.1
![fc_nn](Loesung.jpeg)